src/range.cr

```diff
@@ -342,6 +342,48 @@ struct Range(B, E)
     end
   end
 
+  # Optimized version of `Enumerable#sample` that runs in O(1) time when `self`
+  # is an `Int` or `Float` range. In these cases, this range is considered to be
+  # a distribution of numeric values rather than a collection of elements, and
+  # the method simply calls `random.rand(self)`.
+  #
+  # Raises `ArgumentError` if `self` is an open range.
+  def sample(random = Random::DEFAULT)
+    {% if B == Nil || E == Nil %}
+      {% raise "Can't sample an open range" %}
+    {% end %}
+
+    {% if B < Int && E < Int %}
+      random.rand(self)
+    {% elsif B < Float && E < Float %}
+      random.rand(self)
```

Comment:
Maybe these should better be runtime checks to recognize `Range(Int32?, Int32?).new(1, 2)`.

Reply:
Could be both.

Reply:
The new commit already forwards nilable `#sample` to the non-nilable path, the compile-time checks are needed so that `Range(Float64, Float64)#sample` and its nilable variants never instantiate `super`.